### PR TITLE
Add warning log to see what http connection is having trouble

### PR DESCRIPTION
**Pull request by Mtodiy**
During uploading some file with the HTTP File upload, there was some connection timeout that occurred. Afterward in the logging you are unable to see in the logging to what http connection the timeout occured. 

This pull request just adds one line of logging so you can better see to what http connection specifically is having problems.

### CODE REVIEW

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 1]
```diff
@@ -499,6 +499,9 @@ private void upload(InputStream iStream, long fileSize, Slot slot, UploadProgres
                                 + putUrl);
             }
         }
+        catch (IOException e) {
+            LOGGER.log(Level.WARNING, String.format("Exception occurred while uploading to %s", putUrl.toString()));
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:503]: That would swallow the `IOException` and not throw it up to the caller. That does not appear to be sensible behavior for an library.

Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:503]: You are right. Changed it to a simple catch rethrow, just adding the logging. I Amended the commit

Comment (Flowdalic) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:503]: It is unusual to log an exception and throw it. I assume your motivation is to record `putUrl`, right? That's usually the callers job to log this when he catches the exception. Another option would be to add a specialized subclass of `IOException` that also contains the `putUrl`, say `HttpUploadIoExcetpion`. But I thing that would clutter the API with next to now benefit. But maybe I am misunderstand the problem you are trying to solve here?

Comment (Flowdalic) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:503]: I just noticed that `putUrl` is not known to the caller. In that case, having an specialized exception `HttpUploadException` appears sensible. The exception should include `slot` and `fileSize`, as I think that is all contextual information that is sensible to include.

Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: Indeed, that was the motivation, since putUrl is never known to the caller in some setups it can be hard to debug what server we should investigate if the upload fails. 

I added the sensible custom exception., instead of rethrowing. Let me know what you think, cheers.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 2]
```diff
@@ -494,11 +495,12 @@ private void upload(InputStream iStream, long fileSize, Slot slot, UploadProgres
             case HttpURLConnection.HTTP_NO_CONTENT:
                 break;
             default:
-                throw new IOException("Error response " + status + " from server during file upload: "
-                                + urlConnection.getResponseMessage() + ", file size: " + fileSize + ", put URL: "
-                                + putUrl);
+                throw new HttpUploadException(status, urlConnection.getResponseMessage(), fileSize, putUrl);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:498]: This should be an extra subclass of `HttpUploadException`, which itself, should become an abstract class.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/exception/HttpUploadException.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.IOException;
+import java.net.URL;
+
+import org.jivesoftware.smackx.httpfileupload.element.Slot;
+
+/**
+ * An exception class to provide additional information in case of exceptions during file uploading.
+ *
+ */
+public class HttpUploadException extends IOException {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/exception/HttpUploadException.java:29]: Creating a new Java package for a single class seems like overkill. Just place it in o.j.smackx.httpfileupload. Also, what's the point if the constructors have parameters, but the arguments of the parameters are never accessible by the user? I think this class is missing some `private final` fields and the according getters.

**Review by Flowdalic — CHANGES_REQUESTED** [review]

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/exception/HttpUploadException.java (hunk 3 above)
Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/exception/HttpUploadException.java]: Fair enough, I made the class more accessible with appropriate getters

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java (hunk 2 above)
Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: Made the class abstract, and made two concrete and more specific exceptions

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 4]
```diff
@@ -494,11 +497,12 @@ private void upload(InputStream iStream, long fileSize, Slot slot, UploadProgres
             case HttpURLConnection.HTTP_NO_CONTENT:
                 break;
             default:
-                throw new IOException("Error response " + status + " from server during file upload: "
-                                + urlConnection.getResponseMessage() + ", file size: " + fileSize + ", put URL: "
-                                + putUrl);
+                throw new UnsuccessfulUploadException(status, urlConnection.getResponseMessage(), fileSize, putUrl);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:500]: I am not super happy with the name of the exception, but you know: naming things is hard. But what about `HttpUploadErrorException`, since it contains an HTTP error code?

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 5]
```diff
@@ -494,11 +497,12 @@ private void upload(InputStream iStream, long fileSize, Slot slot, UploadProgres
...
             }
         }
+        catch (IOException e) {
+            throw new UnexpectedUploadException(fileSize, slot, e);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:504]: As above, I would name things exception `HttpUploadIoException`, since it basically wraps an `IOException`.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * Exception thrown when an unexpected exception occured during the upload.
+     */
+    public static class UnexpectedUploadException extends HttpUploadException {
+
+        private static final long serialVersionUID = 5940866318073349451L;
+        private final long fileSize;
+        private final Slot slot;
+
+        public UnexpectedUploadException(long fileSize, Slot slot, Throwable cause) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java:90]: The `cause` parameter here should be of type `IOException`, since this is the only type we catch.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+
+        private static final long serialVersionUID = 8494356028399474995L;
+        private final int httpStatus;
+        private final String responseMsg;
+        private final long fileSize;
+        private final URL putUrl;
+
+        public UnsuccessfulUploadException(int httpStatus, String responseMsg, long fileSize, URL putUrl) {
+            super(String.format("Error response %d from server during file upload %s, file size: %d, put Url: %s",
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java:56]: Note that this is a situation where you typically would not use `String.format()`, as you do not use any format specifiers. String concatenation using the `+` operator would just work fine here, would be more readable and more Java-idomatic.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+     * Exception thrown when an unexpected exception occured during the upload.
+     */
+    public static class UnexpectedUploadException extends HttpUploadException {
+
+        private static final long serialVersionUID = 5940866318073349451L;
+        private final long fileSize;
+        private final Slot slot;
+
+        public UnexpectedUploadException(long fileSize, Slot slot, Throwable cause) {
+            super(String.format("Unexpected error occurred during file upload, file size: %d, put Url: %s",
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java:91]: See my remark about `String.format()` above.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                    fileSize, slot.getPutUrl()), cause);
+            this.fileSize = fileSize;
+            this.slot = slot;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java:94]: I think it would be sensible to also save the causing `IOException` here, so we can provide a getter `IOException getCausingIoException()`, which returns the most narrow type: `IOException` in this case.

**Review by Flowdalic — CHANGES_REQUESTED** [review]
We are getting close. Thanks for you work. Please find my remarks attached. Also note that I suggest that you squash the commits into one at the end.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+import java.net.URL;
+
+import org.jivesoftware.smackx.httpfileupload.element.Slot;
+
+
+/**
+ * An exception class to provide additional information in case of exceptions during file uploading.
+ *
+ */
+public abstract class HttpUploadException extends IOException {
```
[the remark this answers is not in the document]
Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java]: fair enough

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java (hunk 4 above)
Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: Sure, though I liked my functional descriptive ones, this is more factual indeed

**Comment by Mtodiy** [comment]
Alright, made the amends. Looking forward to your feedback ;)

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 11]
```diff
@@ -57,7 +59,6 @@
 import org.jivesoftware.smackx.xdata.packet.DataForm;
 
 import org.jxmpp.jid.DomainBareJid;
-
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:60]: Unrelated whitespace change, please remove this.

**Review by Flowdalic — CHANGES_REQUESTED** [review]
Some minor things left, then it's ready.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/AbstractHttpUploadException.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ */
+public abstract class AbstractHttpUploadException extends IOException {
+
+    private static final long serialVersionUID = 1L;
+    private final long fileSize;
+    private final Slot slot;
+
+    protected AbstractHttpUploadException(long fileSize, Slot slot, String message) {
+        super(message);
```
[the remark this answers is not in the document]
Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/AbstractHttpUploadException.java]: resolved!

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/AbstractHttpUploadException.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
+/**
+ *
+ * Copyright the original author or authors
```
[the remark this answers is not in the document]
Comment (Mtodiy) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/AbstractHttpUploadException.java]: Finally a place where I can legally just scribble my name onto!

**Comment by Mtodiy** [comment]
Put the last changes in, cheers!

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/AbstractHttpUploadException.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
+/**
+ *
+ * Copyright Micha Kurvers
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/AbstractHttpUploadException.java:3]: Please add the year to the copyright statement. See also the other copyright statements in Smack.
